src/monodroid/jni/strings.hh

```diff
@@ -0,0 +1,731 @@
+				return false;
+			}
+
+			return _start[index] == ch;
+		}
+
+		force_inline ssize_t find (const char ch, size_t start_index) const noexcept
+		{
+			if (!can_access (start_index)) {
+				return -1;
+			}
+
+			while (start_index <= _length) {
+				if (_start[start_index] == ch) {
+					return static_cast<ssize_t>(start_index);
+				}
+				start_index++;
+			}
+
+			return -1;
+		}
+
+		template<typename T>
+		force_inline bool to_integer (T &val, size_t start_index = 0, int base = 10) const noexcept
+		{
+			static_assert (std::is_integral_v<T>);
+			constexpr T min = std::numeric_limits<T>::min ();
+			constexpr T max = std::numeric_limits<T>::max ();
+			using integer = typename std::conditional<std::is_signed_v<T>, int64_t, uint64_t>::type;
+
+			if (_length == 0) {
+				return false;
+			}
+
+			if (!can_access (start_index)) {
+				log_error (LOG_DEFAULT, "Cannot convert string to integer, index %u is out of range", start_index);
+				return false;
+			}
+
+			// FIXME: this is less than ideal... we shouldn't need another buffer here
+			size_t slen = _length - start_index;
+			char s[slen + 1];
+
+			memcpy (s, _start + start_index, slen);
+			s[slen] = '\0';
+
+			integer ret;
+			char *endp;
+			bool out_of_range;
+			errno = 0;
+			if constexpr (std::is_signed_v<T>) {
+				ret = strtoll (s, &endp, base);
+				out_of_range = ret < min || ret > max;
+			} else {
+				ret = strtoull (s, &endp, base);
+				out_of_range = ret > max;
+			}
+
+			if (out_of_range || errno == ERANGE) {
+				log_error (LOG_DEFAULT, "Value %s is out of range of this type (%lld..%llu)", s, static_cast<int64_t>(min), static_cast<uint64_t>(max));
+				return false;
+			}
+
+			if (endp == s) {
+				log_error (LOG_DEFAULT, "Value %s does not represent a base %d integer", s, base);
+				return false;
+			}
+
+			if (*endp != '\0') {
+				log_error (LOG_DEFAULT, "Value %s has non-numeric characters at the end", s);
+				return false;
+			}
+
+			val = static_cast<T>(ret);
+			return true;
+		}
+
+	private:
+		force_inline bool can_access (size_t index) const noexcept
+		{
+			if (XA_UNLIKELY (!initialized () || _start == nullptr)) {
+				return false;
+			}
+
+			if (index > _length)
+				return false;
+
+			return true;
+		}
+
+	protected:
+		size_t   _last_index = 0;
+		bool     _fresh = true;
+		char    *_start = nullptr;
+		size_t   _length = 0;
+
+		template<size_t MaxStackSize, typename TStorage, typename TChar> friend class string_base;
+	};
+
+	template<size_t MaxStackSize, bool HasResize, typename T = char>
+	class local_storage
+	{
+		protected:
+		using LocalStoreArray = std::array<T, MaxStackSize>;
+
+	public:
+		static constexpr bool has_resize = HasResize;
+
+	public:
+		explicit local_storage (size_t size) noexcept
+		{
+			static_assert (MaxStackSize > 0, "MaxStackSize must be more than 0");
+			init_store (size < MaxStackSize ? MaxStackSize : size);
+		}
+
+		virtual ~local_storage ()
+		{
+			free_store ();
+		}
+
+		T* get () const noexcept
```

Comment:
I think this should be overloaded based on "`const`-ness":

```cpp
const T* get () const noexcept
{
    return …
}

T *get () noexcept
{
    return …
}
```

The presence of the `const_cast<T*>()` is itself a smell, and wouldn't be needed if `get()` were properly overloaded.

Additionally, as-is there's a potential runtime assert here (unless C++ has changed in this area in the past 20 years), e.g. if you have a global `const local_storage<10, false> v` instance, which is placed by the linker into read-only memory, then `*v.get() = 'a'` will crash.

Reply:
Without the `const` overload the `*v.get() = 'a'` code wouldn't build:

```cpp
const local_storage<20, false> _test (10);
*_test.get () = 'a';
```

would give

```
android-system.cc:222:16: error: read-only variable is not assignable
        *_test.get () = 'a';
        ~~~~~~~~~~~~~ ^
```

I will, however, add the `const` overload as it will come useful elsewhere (and will remove the need for the `const_cast`)

Reply:
I'm not at all sure why we'd *want* the fragment `const local_storage<…> _test(10); *_test.get() = 'a'` to build.  I don't understand why this would be desirable.